addons/crm/models/crm_lead.py

```diff
@@ -286,7 +286,7 @@ def _compute_company_id(self):
                 if lead.team_id and not lead.team_id.company_id and not lead.user_id:
                     proposal = False
                 # no user and no team -> void company and let assignment do its job
-                if not lead.team_id and not lead.user_id:
+                if not lead.team_id and not lead.user_id and not (lead.partner_id and proposal == lead.partner_id.company_id):
```

Comment:
About code: would make a new entry in the "if" list to clearly explain the case, aka something like
```
# no user and no team -> void company and let assignment do its job
if not lead.team_id and not lead.user_id:
    proposal = False
# incompatible with partner's company
if proposal and lead.partner_id.company_id != proposal:
    proposal= False
```